### PR TITLE
ath79: mikrotik: use 64 KiB SPI NOR erase sectors

### PR DESCRIPTION
**Pull request by rogerpueyo**
~~This pull request consists of two commits. The first one removes the CONFIG_MTD_SPI_NOR_USE_4K_SECTORS symbol from the ath79/mikrotik subtarget, so that erase sectors use the default 64 KiB size. Additionally, it makes all soft_config partitions read-only, to avoid accidentally erasing sectors around it when modifying it.~~

This commit removes the CONFIG_MTD_SPI_NOR_USE_4K_SECTORS symbol from the ath79/mikrotik subtarget, so that erase sectors use the default 64 KiB size. The commit message contains details about the rationale behind it.

The approach followed here is the opposite to https://github.com/openwrt/openwrt/pull/4379, which offered terribly slow erase/write performance.

### CODE REVIEW

**Comment by john-tho** [comment]
Hi Roger,

Personally, I would not set the `soft_config` partitions to read only.
The (RouterBOOT sysfs) driver will not allow writing to the soft_config without `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS`. See: https://github.com/openwrt/openwrt/blob/4d0883c011e1f56711f59c2f18315d2cd817dba2/target/linux/generic/files/drivers/platform/mikrotik/rb_softconfig.c#L67

**Comment by rogerpueyo** [comment]
Hi John,

Thank you very much for the comments. I changed the commit message, I hope it is clear and accurate enough.

Also, I double-checked that the read-only property is not needed, and dropped the second commit from the pull request.

**Comment by john-tho** [comment]
Hi Roger,

Looks better to me. Might not need all of it (up to a commiter), but it is accurate now.
May want to detail _Using 4 KiB erase sector size on the sysupgrade images_ (through BLOCKSIZE = 4k),
and that this fixes an open bug, with something like:
Fixes: FS#3492
Fixes: a66eee63368e ("ath79: add mikrotik subtarget")

**Comment by rogerpueyo** [comment]
Yup, it's a huge commit message for a one-line change, but I'd rather keep the rationale behind it in the commit message, just in case the discussion in the PRs is lost.

I added both fixes. :+1:

Thanks!

**Comment by f00b4r0** [comment]
@john-tho Hi John,

> * The OpenWrt patch for [`MTD_SPI_NOR_USE_4K_SECTORS_LIMIT`](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.4/470-mtd-spi-nor-support-limiting-4K-sectors-support-base.patch) may once have worked per-partition. Linux changes mean it now only works per device.
> * Partial eraseblock write can be used with ath79 (Linux 5.4 and lower). It does this by copying the bits that are not requested to be erased from the 64k sector, erasing the whole sector, then restoring the other bits. A kernel bump had that patch very broken for a long time. This was fixed in: [bf2870c](https://github.com/openwrt/openwrt/commit/bf2870c1d9e162f27925e9c40eca9e0069096f3b)

So for my full understanding: moving forward (5.10 and newer) and despite bf2780c, write access to soft_config will not be possible, right? (trying to understand if I should lift the restrictions in the rb_softconfig code).

**Comment by john-tho** [comment]
> So for my full understanding: moving forward (5.10 and newer) and despite bf2780c, write access to soft_config will not be possible, right? (trying to understand if I should lift the restrictions in the rb_softconfig code).

Hey @f00b4r0, correct. For kernel 5.10, there is no OpenWrt patch to do the fake 4K erase as part of a 64K copy, erase and restore. Need to force 4K everywhere with kernel `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS`.

Regarding my mixed erasesizes patch: No advancement: I have not had reply on linux-mtd, or from any of the three OpenWrt & Linux mtdpart.c patch authors I asked separately for any advice or review.

**Comment by adschm** [comment]
This whole subject has been an open issue for a long time. Since nobody from the committer family seemed particularly interested to investigate this properly, I'm going to change the rules of the game now:

I'd be happy if you, the major stakeholders of this platform, find an agreement how we should solve/settle the issue. If the result is generally acceptable, I'd be willing to merge it without looking too closely in order to get this platform somewhat useful (limiting this offer to ath79).

I think of the following people @john-tho @f00b4r0 @rogerpueyo @xback 
Others might be added or join the party of course.

Please consider all side cases, choose wisely and take your time. We've already messed enough and went forwards and backwards with the erase size. I do not intend to change this another time after that (at least for a few _years_).

(And please focus on future support, i.e. master and 5.10+)

**Comment by john-tho** [comment]
Thank you for your interest Adrian,

This turned into a wall of text after looking at the pending 470 patch.

On ath79 Mikrotik NOR devices sysupgrade config restore is broken.
This looks to be due to a mismatch between the image recipe eraseblock size, and the kernel eraseblock size, because `MTD_SPI_NOR_USE_4K_SECTORS_LIMIT` is not working.
This PR is one approach to fixing Mikrotik NOR ath79 sysupgrade config restore, another was https://github.com/openwrt/openwrt/pull/4379.

For uniform erase region (typical) SPI-NOR devices, the kernel selects a single eraseblock size for the NOR device.
With this PR (and broken 470 patch), we no longer [force the kernel to use a 4KiB NOR eraseblock size whenever the SPI-NOR device supports it](https://elixir.bootlin.com/linux/v5.10.68/source/drivers/mtd/spi-nor/core.c#L2575).

Why this happened?
- The OpenWrt patch (pending 470) for `MTD_SPI_NOR_USE_4K_SECTORS_LIMIT` does not work (on [5.4](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.4/470-mtd-spi-nor-support-limiting-4K-sectors-support-base.patch), or [5.10](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.10/470-mtd-spi-nor-support-limiting-4K-sectors-support-base.patch)).
  If it did:
  - ath79 Mikrotik NOR devices would not use a 0x1000 erasesize(note patch 411 below), because their NOR chip size is over the 4MiB limit in the generic config.
  - sysupgrade config restore would work
  - on 5.4 the [OpenWrt pending 411 patch](https://github.com/openwrt/openwrt/blob/master/target/linux/generic/pending-5.4/411-mtd-partial_eraseblock_write.patch) would fake partition-length eraseblock size on writable partitions that are not aligned to the NOR device erasesize.
  - This happened in the kernel jump to 5.4 https://github.com/openwrt/openwrt/commit/4e0c54bc5bc8381e031af5147b66b4dadeecc626#diff-2dd6b0744aaadb9a22422a599b593bbc6f6a6342fadaf1f186bd8de68d4965ba
- Reasoning:
  For uniform erase region SPI-NOR
  `nor->erase_opcode` & `mtd->erasesize` are set by
  ```
  spi_nor_info_init_params
    spi_nor_default_setup
      spi_nor_select_erase
  ```
  These are set before the changes introduced with the 470 patch `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS_LIMIT`, and these values are not modified by it.

  `spi_nor_erase` uses `mtd->erasesize` & `nor->erase_opcode` through `spi_nor_erase_sector`,
so `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS_LIMIT` has no effect (for uniform eraseregions)

If the 470 patch was fixed instead of this PR, ath79 Mikrotik NOR would see: working config restore, but no `soft_config` write above kernel 5.4. Also would expect to find other devices (which use `CONFIG_MTD_SPI_NOR_USE_4K_SECTORS=y`) get a broken sysupgrade config restore because they no longer have a 0x1000 erasesize due to flash device exceeding …`LIMIT`. With review, https://lore.kernel.org/linux-mtd/20210608040719.14431-1-git@johnthomson.fastmail.com.au/ (OpenWrt https://github.com/openwrt/openwrt/pull/3271) should replace the pending 411, 412 & 470 patches, but it got no useful comments on linux-mtd.


Looking at this PR only:
Advantages:
- working sysupgrade config restore
- significantly faster SPI-NOR erase (most noticeable in sysupgrade)
  I tested `time mtd erase firmware` on a different platform and device: OpenWrt 5.10 ramips MT7621 w25q128jv (0xfc0000 partition length). This was 130s with 4KiB eraseblock size, 50s without. Doing 16 times more SPI-NOR erase operations takes a longer time.

Disadvantages:
- no longer have write access to RouterBOOT `soft_config` settings
  - This is already the case for all other Mikrotik platforms in OpenWrt master (mt7621 & ipq40xx)
  - Write access to these settings is not needed on a functioning device
  - There are cases where it is useful: I have a device up a pole, and it is much easier to use RouterBOOT `soft_config` settings to first test an updated initramfs image rather than climb a ladder to retrieve a non-booting device, but I am happy to compile my own image for this.
  - Will need to update wiki to show that if you change soft_config in RouterOS, then move to OpenWrt, you can only change RouterBOOT settings back to default through 5 second reset button hold at boot.

**Comment by f00b4r0** [comment]
My opinion on this matter is that #3271 is the clean & correct approach. It will transparently solve the issue for all mikrotik devices on all archs, and AIUI it also addresses a similar issue with FIS redboot.

Write access to soft_config, while not strictly necessary, can certainly help with setups where the device is not at hands' reach, as pointed out by @john-tho above, and more importantly when RouterOS is no longer installed and serial is not available (the typical use case for OpenWRT), it's the only way to fiddle with the settings (e.g. CPU speed, boot order and boot delay).

In an ideal world, we would get this (#3271) accepted upstream (and I think this should be feasible, more on that below), but if OpenWRT is OK to live with that patch as an out-of-tree extra, then I advocate going that way as the best technical solution over both the existing situation (4K EB on ath79) and this PR. Otherwise, this PR is the way to go.

The patch in #3271 is clear and readable, meaning forward maintainability is much improved. @john-tho would you be ok to maintain that one patch in-tree? You clearly have a good grasp of this code and subsystem.

Re getting this accepted upstream: I think no response to the email(s) is unfortunate but is still not the same as a NACK, and shouldn't be regarded as so. Getting things upstream is hard, especially when they're low priority or corner cases not on anybody's radar.

I think one of the issues is that the current commit message does not make it immediately obvious why the patch is needed and that it's a NOP (i.e. harmless) for non-target hardware (items which are necessary to get noticed in my experience). I would suggest rewording the commit message, cleaning up inconsistencies in the patch (e.g. u32/uint32_t) and sending the patch again, also not as RFC but as an actual patch. Upstream will take it more seriously then and is more likely to reply with comments. I can also try tracking down specific maintainers to directly CC on the next send, so that it gets more attention.

HTH

**Comment by adschm** [comment]
> My opinion on this matter is that #3271 is the clean & correct approach. It will transparently solve the issue for all mikrotik devices on all archs, and AIUI it also addresses a similar issue with FIS redboot.

The problem with #3271 is that it touches stuff in generic that I don't understand. This is an option _I_ cannot offer.

**Comment by john-tho** [comment]
> The problem with #3271 is that it touches stuff in generic that I don't understand. This is an option _I_ cannot offer.

Yes, that is fine, and I support it. I need a committer to understand what it does. #3271 is long term compared to this PR. As-is, it does need the setting that this PR uses (`CONFIG_MTD_SPI_NOR_USE_4K_SECTORS` unset). I will do some more work on and send #3271 to linux-mtd and openwrt-devel once again.

The choice (for this PR) is if we are happy losing RouterBOOT `soft_config` write support in order to get sysupgrade config restore working.
Alternatively, we fix pending 470 (`MTD_SPI_NOR_USE_4K_SECTORS_LIMIT`), or set the image blocksize (#4379)

**Comment by f00b4r0** [comment]
@adschm understood.
@john-tho Agreed. It makes sense to go with this PR then: we restore original performance and sysupgrade at the price of soft config write, which IMO is fine. If later #3271 makes progress it will be pure bonus and this PR won't interfere anyway.

**Comment by rogerpueyo** [comment]
Dear @john-tho @f00b4r0 @adschm @xback all,

First, thanks for taking care of this issue, hopefully this time for good.

I went through your comments and, while some of the technical details are beyond my knowledge, I agree with john-tho's and f00b4r0's arguments: using the 64k block size at the price of not writing soft_config.

Pros:
 - "normal" performance
 - sysupgrade keeps settings
 - doesn't block further work to selectively use 4k/64k (e.g., #3271)
 
 Cons:
  - soft_config read-only (although workarounds exist, like compiling custom image)

> The choice (for this PR) is if we are happy losing RouterBOOT `soft_config` write support in order to get sysupgrade config restore working.

> Agreed. It makes sense to go with this PR then: we restore original performance and sysupgrade at the price of soft config write, which IMO is fine.

I also think that having sysupgrade config restore is more an expected feature than writing to soft_config is. So, let's go for this PR's approach :pray:.
